pyemma/coordinates/transform/transformer.py

```diff
@@ -35,15 +35,89 @@
 __all__ = ['Transformer']
 __author__ = 'noe, marscher'
 
+class TransformerIteratorContext(object):
+
+    def __init__(self, stride=1, lag=0):
+        self._lag = lag
+        self.__init_stride(stride)
+
+    def __init_stride(self, stride):
+        self._stride = stride
+        if isinstance(stride, np.ndarray):
+            keys = stride[:, 0]
+            self._trajectory_keys, self._trajectory_lengths = np.unique(keys, return_counts=True)
+        else:
+            self._trajectory_keys = None
+        self._uniform_stride = TransformerIteratorContext.is_uniform_stride(stride)
+        if not self.uniform_stride and not self.is_stride_sorted():
+            raise ValueError("Currently only sorted arrays allowed for random access")
+
+    def ra_indices_for_traj(self, traj):
+        """
+        Gives the indices for a trajectory file (without changing the order within the trajectory itself).
+        :param traj:
+        :return:
+        """
+        assert not self.uniform_stride, "requested random access indices, but is in uniform stride mode"
+        return self._stride[self._stride[:, 0] == traj][:, 1] if traj in self.traj_keys else np.array([])
```

Comment:
actually you can pretty easily, in your branch. Try to obtain two equal
trajectory objects, one with this sorted-random-access and once with
frames_from_file.py

On 28.07.2015 16:40, Moritz Hoffmann wrote:
> In pyemma/coordinates/transform/transformer.py
> <https://github.com/markovmodel/PyEMMA/pull/458#discussion_r35654764>:
> 
>> +            keys = stride[:, 0]
>> +            self._trajectory_keys, self._trajectory_lengths = np.unique(keys, return_counts=True)
>> +        else:
>> +            self._trajectory_keys = None
>> +        self._uniform_stride = TransformerIteratorContext.is_uniform_stride(stride)
>> +        if not self.uniform_stride and not self.is_stride_sorted():
>> +            raise ValueError("Currently only sorted arrays allowed for random access")
>> +
>> +    def ra_indices_for_traj(self, traj):
>> +        """
>> +        Gives the indices for a trajectory file (without changing the order within the trajectory itself).
>> +        :param traj:
>> +        :return:
>> +        """
>> +        assert not self.uniform_stride, "requested random access indices, but is in uniform stride mode"
>> +        return self._stride[self._stride[:, 0] == traj][:, 1] if traj in self.traj_keys else np.array([])
> 
> It just might be the mdtraj implementation as well (and therefore the
> trajectory file format).. i am happy to run some performance tests.
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/markovmodel/PyEMMA/pull/458/files#r35654764>.
> 

-- 
Dr. Guillermo Pérez-Hernández
Freie Universität Berlin
Institute for Mathematics
Arnimallee 6
D-14195 Berlin
tel       0049 30 838 75776
http://userpage.fu-berlin.de/gph82/
--